spyder_kernels/customize/spydercustomize.py

```diff
@@ -303,16 +308,34 @@ def _patched_preparation_data(name):
 #==============================================================================
 # Pdb adjustments
 #==============================================================================
-class SpyderPdb(pdb.Pdb):
+class SpyderPdb(pdb.Pdb, object):  # Inherits `object` to call super() in PY2
 
     send_initial_notification = True
     starting = True
 
+    def __init__(self, *args, **kwargs):
+        """Init Pdb."""
+        self.continue_if_has_breakpoints = True
+        super(SpyderPdb, self).__init__(*args, **kwargs)
+
     # --- Methods overriden by us
     def preloop(self):
         """Ask Spyder for breakpoints before the first prompt is created."""
-        if self.starting:
-            get_ipython().kernel._ask_spyder_for_breakpoints()
+        try:
+            _frontend_request(blocking=False).set_debug_state(True)
+            if self.starting:
+                breakpoints = _frontend_request().get_breakpoints()
+                self.set_spyder_breakpoints(breakpoints)
+        except (CommError, TimeoutError):
+            logger.debug("Could not get breakpoint from the frontend.")
+
+    def postloop(self):
+        """Notifies spyder that the loop has ended."""
+        try:
+            _frontend_request(blocking=False).set_debug_state(False)
+        except (CommError, TimeoutError):
+            logger.debug("Could not send postloop to frontend.")
```

Comment:
Please change this to

    Could not send debugging state to the frontend.

----

From #112:

> Aldo this could be useful in spyder to make decisions on debugging. Now in the spyder code, _reading is taken as meaning _reading_pdb in many places, but you could have a call to input() where it might not be true. It wold therefore be useful to have a way of telling if debugging is really taking place.

Yes, I totally agree that using `_reading` is quite brittle because a call to `input()` also sets that attribute to True. So it's really nice that we're finally able to differentiate between `_reading` and debugging proper. Thanks a lot for that!

Are you using this in other places in PR spyder-ide/spyder#9343? (I didn't see that).

Reply:
No

Reply:
Ok, no problem. We can add another handler later for Spyder to query the kernel if it's in debugging or not.